src/openrct2/interface/Viewport.cpp

```diff
@@ -669,44 +669,38 @@ void viewport_update_sprite_follow(rct_window* window)
 void viewport_update_smart_sprite_follow(rct_window* window)
 {
     auto entity = TryGetEntity(window->viewport_smart_follow_sprite);
-    if (entity == nullptr)
+    if (entity == nullptr || entity->Type == EntityType::Null)
     {
         window->viewport_smart_follow_sprite = SPRITE_INDEX_NULL;
         window->viewport_target_sprite = SPRITE_INDEX_NULL;
+        return;
     }
-    else if (entity->Type == EntityType::Guest || entity->Type == EntityType::Staff)
+
+    switch (entity->Type)
     {
-        Peep* peep = TryGetEntity<Peep>(window->viewport_smart_follow_sprite);
-        if (peep == nullptr)
+        case EntityType::Vehicle:
+            viewport_update_smart_vehicle_follow(window);
+            break;
+
+        case EntityType::Guest:
+        case EntityType::Staff:
         {
-            // will never happen
-            window->viewport_smart_follow_sprite = SPRITE_INDEX_NULL;
-            window->viewport_target_sprite = SPRITE_INDEX_NULL;
-            return;
+            Peep* peep = static_cast<Peep*>(entity);
+            if (peep->Is<Guest>())
+                viewport_update_smart_guest_follow(window, peep);
+            else if (peep->Is<Staff>())
+                viewport_update_smart_staff_follow(window, peep);
```

Comment:
Can probably be split, no local Peep needed by passing `entity->As<Guest/Staff>()` to the functions.